### ISSUE TITLE
Inconsistency with the new version

### ISSUE THREAD

**Issue by madjoe** [lingtjien/Grid-Tiling-Kwin#92]
The same bug appears every here and there: when I open the System Settings window, sometimes it's in float mode, with the titlebar, but when I restart, it appears in tiling mode without the border, which is expected behavior.

My terminal (konsole) is always tiled, even I put it in Client Ignored by name.

My rambox and telegramdesktop apps are still always tiled, even though I put them in Client Min Space (2). With the older version of your script I was able to get them tiled. Now it seems to be impossible. :(

**Comment by lingtjien** [comment]
A lot of things have changed, including the ignoring, which is now done by Regex. So your old settings will most likely not work with this version. I'm assuming you're using your old settings which had ignored names separated by a `,` but now it's regex (so `|`, and the whole list is included in the settings)

Easiest is if you put the settings to default (bottom left button in the UI) and then change those settings to what you had previously but according to the new format. Could you try that and report back?

**Comment by madjoe** [comment]
Ok, I just tried it (after restarting Manjaro) and something has been fixed, but something's still off:

- System settings windows displays tiled and after a second/third run it starts in floating mode [status: unchanged].
- Konsole now properly displays as tiled [status: fixed].
- My rambox and telegramdesktop are still not tiled [status: unchanged].

My current settings:

```
Rows: 2, 2, 2, 2, 2, 2, 2, 2, 2, 2
Columns: 2, 2, 2, 2, 2, 2, 2, 2, 2, 2

Client Min Space
Name: texstudio, inkscape, krita, gimp, designer, creator, kdenlive, kdevelop, chromium, kate, spotify, rambox, telegramdesktop
Space: 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2

Client Ignored
Name: kwin_wayland|ksmserver|krunner|latte-dock|[Pp]lasma|plugin-container|wine|yakuake|overwatch|kcalc|audacious|deadbeef|atom|eclipse|code|deja-dup|vlc|gwenview|xsane|xsane-gimp|kodi|kodi-standalone|steam|anydesk|konsole|mpv|libreoffice-math
Caption: Trace Bitmap (Shift+Alt+B)|Document Properties (Shift+Ctrl+D)|Media viewer|Utility Window|soffice.bin
```

Not sure about the `Caption` part, since resetting it to the default displays this field empty - do you expect comma separated values, or pipe-separated values. Should it be regex-ready, or I may put the Caption as displayed on screen?

A proper warning message would be pretty helpful here, after validation.

**Comment by madjoe** [comment]
BTW, I've noticed another weird issue that I haven's seen before: if I'm in a full screen and when I press Volume Up/Down keys on my keyboard, the application goes from full screen to tiled!

**Comment by lingtjien** [comment]
> BTW, I've noticed another weird issue that I haven's seen before: if I'm in a full screen and when I press Volume Up/Down keys on my keyboard, the application goes from full screen to tiled!

Oh yeah, a bit too easy on the render calls on my part, try the latest dev branch

**Comment by madjoe** [comment]
> Oh yeah, a bit too easy on the render calls on my part, try the latest dev branch
That did the trick. The issue is gone now.

> TLDR: Both captions and names are now Regex, so separate them by `|` instead of `,`

How about Names for Client Min Space? Resetting the settings to defaults still holds commas over there, making the whole configuration procedure a bit inconsistent.

> The only difference between the two is that captions always get surrounded by `^` before and `$` after, which in Regex land means to match the beginning of the word and the end of the word.
> 
> So captions have to fully match the whole caption, whereas names do not. It's the same Regex matcher as the regular JS one. So if you want to test, you can open any browser and hit `F12` and type `RegExp('insert your setting string here').test('A client caption that is very long to test')` and if it returns `true` then it means that the client caption that you inserted is ignored using the regex that you supplied. (So the only difference between client names and captions is that for captions it is always with the extra beginning `^` and end match `$`, so that would be `RegExp('^insert your setting string here$').test('A client caption that is very long to test')`
> 
> I always use this site to build my regexes https://regexr.com/ and that one is also JS based, I believe so should work exactly the same

Regexr and Developer Tool in Firefox giving me different results as per double escaping the special characters, e.g.:

```
RegExp('^Trace Bitmap \\(Shift\\+Alt\\+B\\)$').test('Trace Bitmap (Shift+Alt+B)')
> true
```
The same RegEx will be `/^Trace Bitmap \(Shift\+Alt\+B\)$/` in Regexr. Should I really take care of escaping the special chars? Seems like an overkill for an average user.

> But you can probably remove the ignored captions that you previously used, because the new way (which also works with wayland) is to ignore based on `transient` and those captions that I previously ignored are `transient` clients anyways, so they're already ignored by default now anyways. (Like the `Trace Bitmap (Shift+Alt+B)` was a dialog window of inkscape that didn't properly set itself as a `dialog` type, so I needed a caption ignore to not tile that client, but since it's a child of the inkscape window it is `transient` so that means it's ignored now by default, so you don't need that caption ignore any longer)

Ok, I will try to remove the _Caption_ part altogether as per your recommendation and will report back if I find anything funky again.

**Comment by lingtjien** [comment]
> How about Names for Client Min Space? Resetting the settings to defaults still holds commas over there, making the whole configuration procedure a bit inconsistent.

I want to do that based on Regex as well, but I'm still debating on how to do it, because you need a regex for every min space. so you need to give the user the ability to input multiple regexes. But yeah for now that one is the same as before and uses a comma separated list

**Comment by madjoe** [comment]
> True, I agree that creating regexes is a bit of a steep learning curve for an average user, but I don't think an average user now really needs the ignore caption feature anyways, since the feature was made to fix a problem that has now been solved.

Ok, since you already did lots of work to automatically ignore certain sub-windows, maybe you may display this in a more intuitive fashion, i.e. separating Basic from Advanced settings with Toggle hide/show, similar to Settings within the Chromium browser (currently by putting only `Ignore Caption` in Advanced settings area)?

> The system settings application likes to set its own geometry at startup... which results in it now showing up as tiled a `Grid-Tiling: Refresh` should fix that. Btw, KDE now also offers a setting to not let applications remember their geometry, see the bottom of the `README.md`

I just unchecking `Allow KDE apps to remember the positions of their own windows`, but it didn't help: telegramdesktop and rambox windows are clearly floating and `Meta`+`R` doesn't change anything.

> Are they not tiled correctly at startup or are they ignored completely and unable to tile, even after a `Grid-Tiling: Refresh`?

Those two apps are completely ignored and unable to tile on startup, even after `Grid-Tiling: Refresh`.

**Comment by lingtjien** [comment]
> separating Basic from Advanced settings with Toggle hide/show

I'm refactoring the whole UI anyways and because I wanted to add per desktop grid sizing, which is now in dev, but I don't think the ignore caption settings is unintuitive now that every name matcher is done by regex.

> Those two apps are completely ignored and unable to tile on startup

Hmm, weird, sounds like they're ignored. I just installed `telegram-desktop` and it works perfectly fine for me, tiles correctly. Could you please try with the default settings and see if it tiles? (So we can conclude if it's due to your specific settings or something else)

**Comment by madjoe** [comment]
> Hmm, weird, sounds like they're ignored. I just installed `telegram-desktop` and it works perfectly fine for me, tiles correctly. Could you please try with the default settings and see if it tiles? (So we can conclude if it's due to your specific settings or something else)

I did reset the settings to its default, restarted my OS and opened my apps again: same behavior, nothing has changed. Those two apps are opened in floating mode and can't get them back to grid.

Not sure if the following may bother your script; I have these settings for BOTH of those apps in my Window management (KDE Plasma), but it used to work with an older version just fine. Now it doesn't work:

![Settings](https://i.imgur.com/ylMARCj.png)

**Comment by lingtjien** [comment]
I can reproduce it with the `Skip taskbar` `Force` rule, but it's not related to `telegram-desktop`, I can reproduce it for every client, interesting... I'll take a look

Could you try to remove that rule and see if it works for you? (So that I'm sure we're dealing with the same behavior)

**Comment by madjoe** [comment]
> I can reproduce it with the `Skip taskbar` `Force` rule. Could you try to remove that rule and see if it works for you? (So that I'm sure we're dealing with the same behavior).

Good catch. I could confirm this behavior when `Skip taskbar` is forced for any window.

> The only way to properly solve it would be to request a method in the KWin API to allow scripting to send clients to a specific activity...

Agreed. This should be discussed with the KWin API crew.

> A way to hack it, if you don't use activities is to change line `175` in `Manager.qml` to `<=` instead of `===`, this also makes the script work for wayland since that's the thing that blocks wayland and makes all clients ignored because they're on all activities in wayland, because wayland doesn't support activities yet

I don't use activities, but couldn't this be automated temporarily? E.g. the script checks if activities has been used and then uses `<=` instead `===` by default? I'm sure activities are rarely used anyway.

**Comment by madjoe** [comment]
Unfortunately, this `<=` hack in line 175 opens another issue:
When I apply this change, then open `rambox` and `telegram-desktop`, the windows are tiled properly. But as soon as I click on an image within the `telegram-desktop`, the image is tiled as well (!), but without this "hack" it has been opened properly as float. It seems that this "hack" also messes with your automation for ignoring the sub-windows.

**Comment by lingtjien** [comment]
And `rambox` works fine now?

So it's only the opening of an image in `telegram` that opens it in a new tiled window that was previously not tiled (so a different issue)?

Oh yeah I probably should lol

**Comment by madjoe** [comment]
> And `rambox` works fine now?

No, they both have the same behavior. If I apply the `<=` hack, then all apps work fine, but it seems like Clients Ignored doesn't work at all anymore. If I revert back the line 175 to `===`, then the tiling doesn't work, and Clients Ignored works just fine. Previously I thought that the hack `<=` affects only the automation for ignoring the sub-windows, but it seems it ignores all settings for Clients Ignored altogether, including your automation.

> So it's only the opening of an image in `telegram` that opens it in a new tiled window that was previously not tiled (so a different issue)?

No, it happens for all the applications the same way, not just for `telegram`. 
Yes, this is a new issue I guess.

> If you remove the KWin custom rules you made for telegram, does it float or tile then?

Let me test this... I will update you in a bit.

**Comment by lingtjien** [comment]
If I take the latest `develop` branch implementation and change line `175` to `<=` and I add `|konsole` to the client ignored list, then it ignores konsole and everything else is still tiled normally...

But can't you just revert the change in line `175` and then remove that KWin rule `skip taskbar force`? I mean... that rule puts the client on all activities and I don't see how one could tile a client logically across all activities (with 1 geomtry) when it's active on more than 1 activity...

**Comment by madjoe** [comment]
> If I take the latest `develop` branch implementation and change line `175` to `<=` and I add `|konsole` to the ignored name list, then it ignores konsole and everything else is still tiled normally...

Ah, you're right. It's just you've changed the settings UI again with the new dev and forced reset settings to defaults, that's why my settings were dismissed. I pressed "Default" and set my custom settings manually, now I can see my terminal and calculator floating.

> But can't you just revert the change in line `175` and then remove that KWin rule `skip taskbar force`? I mean... that rule puts the client on all activities and I don't see how one could tile a client logically across all activities (with 1 geomtry) when it's active on more than 1 activity...

The thing is I already many apps in my panel and would like to narrow it down to only necessary apps. Since chat apps are resident in System tray, I don't see any reason to hold them at the panel as well. I may do that, but it's just not what I want. I think you were right yesterday - multi-activity tiling should be discussed with KWin API developers, I guess.

BTW, I tested removing all my KWin custom Window rules and behavior is just the same with `<=` code (with my custom rules or without it).

**Comment by lingtjien** [comment]
> It's just you've changed the settings UI again

Oh yeah I should have mentioned that, my bad. I'm kinda busy refactoring the whole UI and adding some new stuff...

**Comment by lingtjien** [comment]
@madjoe I added a new delay setting. It allows you to specify how many `ms` the script should wait before tiling new clients. Since you had problems with BleachBit not tiling correctly at startup. Could you give it a try and maybe increase the delay time to see if it works, the max value is `9999`, but I don't think anyone would want to wait 10 seconds before new clients are tiled, everything works with a default time of `0` for me, but I'm curious to your experience with applications you previously had trouble with and see if increasing this time fixes it.

**Comment by madjoe** [comment]
> Could you give it a try and maybe increase the delay time to see if it works, the max value is `9999`, but I don't think anyone would want to wait 10 seconds before new clients are tiled, everything works with a default time of `0` for me, but I'm curious to your experience with applications you previously had trouble with and see if increasing this time fixes it.

Interestingly, I pulled the new develop branch (I didn't change `===` to `<=`), restarted Manjaro and ran `bleachbit`. It worked immediately just fine, without any additional gap like it previously did. Well done!

I really hope the multi-activity tiling could be resolved somehow, since that's the only feature missing with this beautiful script. Everything else works just fine now, thanks! :)

**Comment by lingtjien** [comment]
GIve the dev branch another try

**Comment by madjoe** [comment]
Almost there. Now I can have my `rambox` and `telegram` apps tiled (with custom Window rules enabled), but it seems to have issues when working in multiple workplaces. For example, those chat apps should open in my workspace #6. Then I go to workspace #3 and I try to open another application and it opens in workspace #4 by leaving workspace #3 completely empty.

In another try, I try to open `bleachbit` in workspace #3, but then `rambox` jumps to workspace #3 and `telegram` moves to workspace #4. I couldn't find the pattern, but the tiling seems to work now, only the positioning of windows in workspaces are messed up in some circumstances for some reason.

I also noticed a certain delay before rendering the window. Could you explain the main idea behind, maybe I could help to debug it properly?

**Comment by lingtjien** [comment]
give the dev branch another try

**Comment by madjoe** [comment]
I tested it pretty extensively today and it's still not there. Now every time I restart my OS, rambox and telegram launches automatically and when I open each app from systray they open in floating mode. Then I need to exit from both apps and run them again, then the first app starts tiled, the second one not, and after a few more tries of hide/show window, they get tiled properly, but after all the process I've mentioned.

Then I tested the Media viewer window inside telegram and it starts tiled (which is not expected) and then I close it, open it again a few times and then it opens as float (in a full screen, as expected).

There was a few times when I opened several applications and the rambox jumped to the same workspace all of a sudden for no obvious reason.

**Comment by lingtjien** [comment]
I'm going to need a few more reproducible steps... in order to debug that... so... trying to digest that info out of your story you did the following?

1. set application to auto launch at startup
2. reboot
3. application looks floating

Is that correct?

I just tried with konsole, put that in the autostart list and rebooted and yes konsole looks like it starts floating, but as soon as I start another instance of konsole or use the `Grid-Tiling: Refresh` it becomes tiled again

Regarding the media viewer of telegram, I'm assuming its `transient` property is `false` but only one way to know for sure
1. open terminal
2. `kdebugsettings`
3. `KWin Scripting` to `Full Debug`
4. open `Manager.qml` add `if (client) {print(client.resourceName, client.caption, client.transient);}` after line `function add(client) {` `179`
5. open the application that you would like to see the property of (telegram image window in this case)
6. `kwin_x11 --replace &`
7. output?

**Comment by madjoe** [comment]
> Is that correct?

Yes.

> output?

```
QApplication: invalid style override 'gtk' passed, ignoring it.
        Available styles: Breeze, kvantum-dark, kvantum, qt5ct-style, Windows, Fusion
qt.svg: link #path2410 is undefined!
Module 'org.kde.kwin.decoration' does not contain a module identifier directive - it cannot be protected from external registrations.
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
file:///usr/share/kwin/aurorae/MenuButton.qml:11: TypeError: Cannot read property 'closeOnDoubleClickOnMenu' of null
QPainter::begin: Paint device returned engine == 0, type: 3
OpenGL vendor string:                   NVIDIA Corporation
OpenGL renderer string:                 Quadro P2000/PCIe/SSE2
OpenGL version string:                  3.1.0 NVIDIA 450.80.02
OpenGL shading language version string: 1.40 NVIDIA via Cg compiler
Driver:                                 NVIDIA
Driver version:                         450.80.2
GPU class:                              Unknown
OpenGL version:                         3.1
GLSL version:                           1.40
X server version:                       1.20.10
Linux kernel version:                   5.4.80
Requires strict binding:                no
GLSL shaders:                           yes
Texture NPOT support:                   yes
Virtual Machine:                        no
The X11 connection broke: I/O error (code 1)
XIO:  fatal IO error 22 (Invalid argument) on X server ":0"
      after 2915 requests (2915 known processed) with 0 events remaining.
The X11 connection broke: I/O error (code 1)
XIO:  fatal IO error 22 (Invalid argument) on X server ":0"
      after 2915 requests (2915 known processed) with 0 events remaining.
Application::crashHandler() called with signal 11; recent crashes: 1
malloc_consolidate(): unaligned fastbin chunk detected
KCrash: crashing... crashRecursionCounter = 2
KCrash: Application Name = kwin_x11 path = /usr/bin pid = 53226
KCrash: Arguments: /usr/bin/kwin_x11 --replace
```

**Comment by lingtjien** [comment]
> Yes. And I can also confirm both windows get tiled after Meta + R.

Alright, so then it's just a startup thing. The screen area hasn't fully completed yet when the startup application is launched so that's why you need an additional call to the render method. One way to solve this would be to delay starting of your autostart applications by a few seconds, so that the whole desktop environment has finished

I could also repurpose the current delay setting for this and use that setting to delay the script before starting. Which is... hmm... I don't know what I think about that...

> ````
> qml: telegram-desktop Telegram false
> qml: kdebugsettings KDebugSettings true
> qml: telegram-desktop Media viewer false
> ````

Yeah, so my suspicion was correct, the media viewer of telegram is set as a main application and not as a child of the `telegram-desktop` application. I can't really solve that... so this is where `ignored captions` come in handy, if you add the string that matches this media viewer then it should be ignored, so in this case `Media viewer` should work, since that's the caption

**Comment by madjoe** [comment]
> One way to solve this would be to delay starting of your autostart applications by a few seconds, so that the whole desktop environment has finished

I just tried add to front of each autostart command `sleep 20 &&` and they did show up with a huge 20s delay, but unfortunately the result is still the same - when I opened the apps, the apps ran as float, not as tiled.

> I can't really solve that (that's either a bug in KWin of not setting the `transient` property correctly, but I highly doubt that, it's more likely something the telegram developers should fix, since the media viewer of the main desktop application should be a child and thus `transient`)

Maybe I could report that to the Telegram team.

> So this is where `ignored captions` come in handy, if you add the string that matches this media viewer then it should be ignored, so in this case `Media viewer` should work, since that's the caption (and if you need to add more in the future separate them by `|`, such as `Media viewer|Some other caption`)

I added `Media viewer` to the Ignored Captions and now it seems to work. I did NOT put any escape character in front of the space, I hope that's how I should proceed when stringing the captions.

**Comment by lingtjien** [comment]
> when I opened the apps, the apps ran as float, not as tiled

Err wut? So the application always opens as float? So it's not related to autostarting?

I just tested
1. create a `test.sh` script with content
```
#!/bin/bash
sleep 10
konsole &
```
2. put that script in autostart
3. reboot
4. konsole is autostarted after 10 seconds and tiled correctly

So delaying the start of autostart applications works fine for me

> I did NOT put any escape character in front of the space

Spaces don't have to be escaped in neither `JS` nor inside regex, you can if you want too in `JS`, `Media viewer` and `Media\ viewer` should give the same result. (If you test that in any browser `F12`: `console.log('a b')` gives the same as `console.log('a\ b')`

**Comment by madjoe** [comment]
Yep. It's not autostart. Even if I close the app (kill the process), then open it again, it behaves the same way. Maybe the previous issue with skipping the taksbar persists?

**Comment by lingtjien** [comment]
With which apps do you have this? And do they have special window rules?

**Comment by madjoe** [comment]
`rambox` & `telegram`

> do they have special window rules?

Yes, they do: https://github.com/lingtjien/Grid-Tiling-Kwin/issues/92#issuecomment-748508939

**Comment by lingtjien** [comment]
Oh okay, so it's only those two. What if you increase the delay in the settings?

**Comment by madjoe** [comment]
With the delay value of `100` it behaves absolutely the same.

**Comment by lingtjien** [comment]
that's `100ms`, try `1000` and see if that does anything

**Comment by lingtjien** [comment]
oh seems to be the initalize minimized setting hmm

**Comment by madjoe** [comment]
I think it works... I just tried it with `4500` and it works, but obviously such value is too big for the delay, especially since it affects all the new opened windows. I need to find a sweet spot and bring it down to the minimal efficient value.

Although, it would be perfect to filter out such delays only for certain windows that really need it. The rest of the windows obviously don't need such delay.

**Comment by lingtjien** [comment]
Yeah, seems to be a different issue. I'll check it out

**Comment by madjoe** [comment]
I just found a setting that may be important for this:

![](https://i.imgur.com/Ejjrh37.png)

**Comment by lingtjien** [comment]
I already found the problem, but it's a bit more rooted so to say...

**Comment by madjoe** [comment]
Maybe this will help... it works with the following scenario even with no delay right after restart:

1. Show `rambox` window (floating)
2. Open any app that should be tiled, e.g. `kate`
3. Kate gets tiled
4. Close `kate` and hide `rambox` window (it stays resident in memory)
5. Show `rambox` (tiled)
6. Show `telegram` (tiled)

Without opening an app that normally will get tiled in between, it doesn't work.

**Comment by lingtjien** [comment]
Try the dev branch

**Comment by madjoe** [comment]
Wow. It works now! :) Good job.

**Comment by madjoe** [comment]
The latest patch seems to fix even some other issues I didn't report here.

**Comment by lingtjien** [comment]
Well, you did report a lot of issues, to be fair 😄 

But yeah, most of these things were already broken previously... because... well... I took a lot of the code from the old stuff...

So yeah... I'm probably going to tag this one 1.1 and then refactor some stuff later again because this last bug made me realize, I made an assumption more than a year ago that I used for the design, that isn't always true apparently... so yeah...

**Comment by lingtjien** [comment]
merged to master
